graphix/pauli.py

```diff
@@ -301,142 +55,104 @@ class Pauli:
+            return f"-{self.symbol}"
+        if self.unit == ComplexUnit.PLUS_J:
+            return f"1j * {self.symbol}"
+        if self.unit == ComplexUnit.MINUS_J:
+            return f"-1j * {self.symbol}"
+        typing_extensions.assert_never(self.unit)
 
     def __matmul__(self, other: Pauli) -> Pauli:
         """Return the product of two Paulis."""
         if isinstance(other, Pauli):
-            if self.__symbol == IXYZ.I:
-                symbol = other.__symbol
-                unit = 1
-            elif other.__symbol == IXYZ.I:
-                symbol = self.__symbol
-                unit = 1
-            elif self.__symbol == other.__symbol:
-                symbol = IXYZ.I
-                unit = 1
-            elif (self.__symbol.value + 1) % 3 == other.__symbol.value:
-                symbol = IXYZ((self.__symbol.value + 2) % 3)
-                unit = 1j
-            else:
-                symbol = IXYZ((self.__symbol.value + 1) % 3)
-                unit = -1j
-            return get(symbol, unit * self.__unit * other.__unit)
+            return _matmul_impl(self.symbol, other.symbol) * (self.unit * other.unit)
         return NotImplemented
 
-    def __rmul__(self, other: ComplexUnit) -> Pauli:
+    def __mul__(self, other: ComplexUnit | SupportsComplexCtor) -> Pauli:
         """Return the product of two Paulis."""
-        if isinstance(other, ComplexUnit):
-            return get(self.__symbol, other * self.__unit)
+        if u := ComplexUnit.try_from(other):
+            return dataclasses.replace(self, unit=self.unit * u)
         return NotImplemented
 
+    def __rmul__(self, other: ComplexUnit | SupportsComplexCtor) -> Pauli:
+        """Return the product of two Paulis."""
+        return self.__mul__(other)
+
     def __neg__(self) -> Pauli:
         """Return the opposite."""
-        return get(self.__symbol, -self.__unit)
-
-
-TABLE = tuple(
-    tuple(tuple(Pauli(symbol, COMPLEX_UNITS[sign][is_imag]) for is_imag in (False, True)) for sign in (False, True))
-    for symbol in (IXYZ.I, IXYZ.X, IXYZ.Y, IXYZ.Z)
-)
-
-
-LIST = tuple(pauli for sign_im_list in TABLE for im_list in sign_im_list for pauli in im_list)
-
-
-def get(symbol: IXYZ, unit: ComplexUnit) -> Pauli:
-    """Return the Pauli gate with given symbol and unit."""
-    return TABLE[symbol.value + 1][unit.sign == Sign.Minus][unit.is_imag]
-
-
-# TODO: Include in Pauli namespace
-I = get(IXYZ.I, UNIT)
-X = get(IXYZ.X, UNIT)
-Y = get(IXYZ.Y, UNIT)
-Z = get(IXYZ.Z, UNIT)
-
-
-def parse(name: str) -> Pauli:
-    """Return the Pauli gate with the given name (limited to "I", "X", "Y" and "Z")."""
-    return get(IXYZ[name], UNIT)
-
-
-def is_int(value: Number) -> bool:
-    """Return `True` if `value` is an integer, `False` otherwise."""
-    return value == int(value)
+        return dataclasses.replace(self, unit=-self.unit)
 
+    @staticmethod
+    def iterate(include_unit: bool = True) -> Iterator[Pauli]:
```

Comment:
why not __iter__()?

Reply:
Because `iter` cannot have any arguments.

Reply:
Anyway I noticed that we may also need to add `__iter__`.

Reply:
Now the class itself is`Iterable`!
Please check a8160e7f06645030b6f4a0b6effd0882ddbc6bd0 .